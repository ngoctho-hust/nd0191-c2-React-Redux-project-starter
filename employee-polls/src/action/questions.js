import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { addUserAnswer } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ANSWER_QUESTION = "ANSWER_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

export function answerQuestion({ authedUser, qid, answer }) {
  return {
    type: ANSWER_QUESTION,
    authedUser,
    qid,
    answer,
  };
}

export function handleAnswerQuestion(info) {
  return (dispatch) => {
    return saveQuestionAnswer(info)
      .then(() => {
        dispatch(answerQuestion(info));
        dispatch(addUserAnswer(info));
      })
      .catch(() => {
        alert("Error when save question answer! Try again!");
      });
  };
}

export function handleAddQuestion(info) {
  return (dispatch) => {
    return saveQuestion(info)
      .then((question) => {
        dispatch(addQuestion(question));
      })
      .catch(() => {
        alert("Error when save question! Try again!");
      });
  };
}
