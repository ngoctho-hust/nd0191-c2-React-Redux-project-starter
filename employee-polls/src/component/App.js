import { useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../action/shared";
import Dashboard from "./Dashboard";
import Header from "./Header";
import { Route, Routes } from "react-router-dom";
import AuthedRoute from "./AuthedRoute";
import Login from "./Login";
import Poll from "./Poll";
import { LoadingBar } from "react-redux-loading-bar";
import PollCreation from "./PollCreation";

const App = (props) => {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, []);

  return (
    <main>
      {props.authedUser && <Header />}
      <LoadingBar />
      {props.loading === true ? null : (
        <Routes>
          <Route path="/login" exact element={<Login />} />
          <Route
            path="/"
            element={
              <AuthedRoute>
                <Dashboard />
              </AuthedRoute>
            }
          />
          <Route
            path="/questions/:id"
            element={
              <AuthedRoute>
                <Poll />
              </AuthedRoute>
            }
          />
          <Route
            path="/new"
            element={
              <AuthedRoute>
                <PollCreation />
              </AuthedRoute>
            }
          />
        </Routes>
      )}
    </main>
  );
};

const mapStateToProps = ({ authedUser, users }) => ({
  authedUser,
  loading: users === null,
});

export default connect(mapStateToProps)(App);
