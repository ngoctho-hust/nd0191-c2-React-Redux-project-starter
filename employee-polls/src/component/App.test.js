import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducer from "../reducer";
import middleware from "../middleware";
import { BrowserRouter } from "react-router-dom";
import { setAuthedUser } from "../action/authedUser";

const store = createStore(reducer, middleware);

describe("App", () => {
  it("should render App component and match snapshot", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  it("should show require login when not login", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const heading = screen.getByText("Log In");
    expect(heading).toBeInTheDocument();
  });

  it("should show Dashboard after logged in", () => {
    store.dispatch(setAuthedUser("sarahedo"));

    render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    const heading = screen.getByText("New Questions");
    expect(heading).toBeInTheDocument();
  });
});
