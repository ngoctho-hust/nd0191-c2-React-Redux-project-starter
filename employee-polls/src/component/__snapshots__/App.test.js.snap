// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App component and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main>
        <div />
        <main>
          <div
            class="container text-center"
          >
            <div
              class="row"
            >
              <div
                class="col-sm"
              >
                <h1>
                  Employee Polls
                </h1>
                <img
                  alt="Employee Polls"
                  class="rounded"
                  src="people.png"
                />
                <h2>
                  Log In
                </h2>
                <form>
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      for="user"
                    >
                      User
                    </label>
                    <input
                      class="form-control"
                      id="user"
                      placeholder="User"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="mb-3"
                  >
                    <label
                      class="form-label"
                      for="password"
                    >
                      Password
                    </label>
                    <input
                      class="form-control"
                      id="exampleInputPassword1"
                      placeholder="Password"
                      type="password"
                      value=""
                    />
                  </div>
                  <button
                    class="btn btn-primary"
                    type="submit"
                  >
                    Submit
                  </button>
                </form>
              </div>
            </div>
          </div>
        </main>
      </main>
    </div>
  </body>,
  "container": <div>
    <main>
      <div />
      <main>
        <div
          class="container text-center"
        >
          <div
            class="row"
          >
            <div
              class="col-sm"
            >
              <h1>
                Employee Polls
              </h1>
              <img
                alt="Employee Polls"
                class="rounded"
                src="people.png"
              />
              <h2>
                Log In
              </h2>
              <form>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="user"
                  >
                    User
                  </label>
                  <input
                    class="form-control"
                    id="user"
                    placeholder="User"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="password"
                  >
                    Password
                  </label>
                  <input
                    class="form-control"
                    id="exampleInputPassword1"
                    placeholder="Password"
                    type="password"
                    value=""
                  />
                </div>
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Submit
                </button>
              </form>
            </div>
          </div>
        </div>
      </main>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
